File: tests/st/ops/test_ops_argmin_ext.py
# Copyright 2024 Huawei Technologies Co., Ltd
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ============================================================================
import pytest
import numpy as np
import mindspore as ms
from mindspore import ops, Tensor, mint, jit, JitConfig
from tests.st.ops.dynamic_shape.test_op_utils import TEST_OP
from tests.st.utils import test_utils
from tests.mark_utils import arg_mark


def generate_random_input(shape, dtype):
    return np.random.randn(*shape).astype(dtype)


def generate_expect_forward_output(x, dim=None, keepdim=False):
    return np.argmin(x, axis=dim)


@test_utils.run_with_cell
def argmin_ext_forward_func(x, dim=None, keepdim=False):
    return mint.argmin(x, dim=dim, keepdim=keepdim)


@test_utils.run_with_cell
def argmin_ext_backward_func(x, dim=None, keepdim=False):
    return ops.grad(argmin_ext_forward_func)(x, dim, keepdim)

def GenInputData(np_data_type, shape=(3, 4, 5)):
    """GenInputData"""
    size = 1
    for s in shape:
        size *= s
    data = np.arange(size).reshape(*shape).astype(np_data_type)
    return Tensor(data)


@arg_mark(plat_marks=['platform_ascend', 'platform_ascend910b'], level_mark='level0',
          card_mark='onecard', essential_mark='essential')
@pytest.mark.parametrize('mode', ['pynative', 'KBK'])
def test_ops_argmin_ext(mode):
    """
    Feature: pyboost function.
    Description: test function argmin forward.
    Expectation: expect correct result.
    """
    input_tensor_list = [
        generate_random_input((2, 3, 4, 5), np.float32)
    ]
    dim_list = [0]
    keepdim_list = [False]

    for i in range(len(input_tensor_list)):
        x = input_tensor_list[i]
        dim = dim_list[i]
        keepdim = keepdim_list[i]
        if mode == 'pynative':
            ms.context.set_context(mode=ms.PYNATIVE_MODE)
            output = argmin_ext_forward_func(ms.Tensor(x), dim, keepdim)
            out_grad = argmin_ext_backward_func(ms.Tensor(x), dim, keepdim)
        elif mode == 'KBK':
            output = (jit(argmin_ext_forward_func, jit_config=JitConfig(jit_level="O0")))(ms.Tensor(x), dim, keepdim)
            out_grad = (jit(argmin_ext_backward_func, jit_config=JitConfig(jit_level="O0")))(ms.Tensor(x), dim, keepdim)
        else:
            output = (jit(argmin_ext_forward_func, jit_config=JitConfig(jit_level="O2")))(ms.Tensor(x), dim, keepdim)
            out_grad = (jit(argmin_ext_backward_func, jit_config=JitConfig(jit_level="O2")))(ms.Tensor(x), dim, keepdim)
        expect = generate_expect_forward_output(x, dim, keepdim)
        np.testing.assert_allclose(output.asnumpy(), expect, rtol=1e-3)
        np.testing.assert_allclose(out_grad.asnumpy(), 0, rtol=1e-3)


@arg_mark(plat_marks=['platform_ascend', 'platform_ascend910b'], level_mark='level1',
          card_mark='onecard', essential_mark='unessential')
def test_argmax_ext_dynamic_shape():
    """
    Feature: Test argmin with dynamic shape in pynative mode and KBK mode.
    Description: call mint.argmin with valid input, dim and keepdim.
    Expectation: return the correct value.
    """
    ms_data1 = ms.Tensor(generate_random_input((2, 3, 4, 5), np.float32))
    dim1 = 0
    keepdim1 = True

    ms_data2 = ms.Tensor(generate_random_input((5, 8, 7), np.float32))
    dim2 = 1
    keepdim2 = False
    TEST_OP(argmin_ext_forward_func, [[ms_data1, dim1, keepdim1], [ms_data2, dim2, keepdim2]],
            'argmin_ext', disable_mode=['GRAPH_MODE'])

File: tests/ut/cpp/ops/test_ops_argmin_ext.cc
/**
 * Copyright 2024 Huawei Technologies Co., Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include <vector>
#include "ops/test_ops_cmp_utils.h"
#include "ir/dtype/number.h"
#include "ops/ops_func_impl/argmin_ext.h"
#include "ops/test_value_utils.h"
#include "abstract/dshape.h"
#include "ops/auto_generate/gen_ops_primitive.h"
#include "utils/tensor_construct_utils.h"

namespace mindspore {
namespace ops {

struct ArgMinExtParams {
  ShapeVector input_shape;
  TypePtr     input_dtype;
  ValuePtr    dim;
  ValuePtr    keepdim;
  ShapeVector output_shape;
  TypePtr     output_dtype;
};

class TestArgMinExt : public TestOps, public testing::WithParamInterface<ArgMinExtParams> {};

TEST_P(TestArgMinExt, dyn_shape) {
  const auto &param = GetParam();
  auto x = std::make_shared<abstract::AbstractTensor>(param.input_dtype, param.input_shape);
  ASSERT_NE(x, nullptr);
  auto dim = param.dim->ToAbstract();
  ASSERT_NE(dim, nullptr);
  auto keepdim = param.keepdim->ToAbstract();
  ASSERT_NE(keepdim, nullptr);

  auto expect_shape = std::make_shared<abstract::Shape>(param.output_shape);
  auto expect_type = std::make_shared<TensorType>(param.output_dtype);
  DoFuncImplInferAndCompare<ArgMinExtFuncImpl>(kNameArgMinExt, {x, dim, keepdim}, expect_shape, expect_type);
}

static std::vector<ArgMinExtParams> GetCases() {
  auto dyn_rank = abstract::TensorShape::kShapeRankAny;
  auto dyn_dim = abstract::TensorShape::kShapeDimAny;
  std::vector<ArgMinExtParams> cases = {
    ArgMinExtParams{{4, 2, 3}, kFloat16, CreateScalar<int64_t>(1), CreateScalar(false), {4, 3}, kInt64},
    ArgMinExtParams{{dyn_rank}, kFloat16, CreateScalar<int64_t>(1), CreateScalar(false), {dyn_rank}, kInt64},
    ArgMinExtParams{{4, 2, 3}, kFloat16, CreateScalar(kValueAny), CreateScalar(false), {dyn_dim, dyn_dim}, kInt64},
    ArgMinExtParams{{4, dyn_dim, 3}, kFloat16, CreateScalar<int64_t>(1), CreateScalar(false), {4, 3}, kInt64},
  };
  return cases;
}

class TestArgMinExtSimple : public TestOps, public testing::WithParamInterface<ArgMinExtParams> {};

TEST_P(TestArgMinExtSimple, simple_infer) {
  const auto &param = GetParam();
  auto x = std::make_shared<tensor::BaseTensor>(param.input_dtype->type_id(), param.input_shape);
  auto dim = param.dim->ToAbstract();
  auto keepdim = param.keepdim->ToAbstract();

  auto expect_shape = std::make_shared<abstract::Shape>(param.output_shape);
  auto expect_type = std::make_shared<TensorType>(param.output_dtype);

  DoFuncImplInferAndCompare<ArgMinExtFuncImpl>(kNameArgMinExt, {x->ToAbstract(), dim, keepdim}, expect_shape, expect_type);
}

INSTANTIATE_TEST_CASE_P(TestArgMinExt, TestArgMinExt, testing::ValuesIn(GetCases()));

INSTANTIATE_TEST_CASE_P(
  TestArgMinExtSimple, TestArgMinExtSimple,
  testing::Values(
    ArgMinExtParams{{4, 2, 3}, kFloat16, CreateScalar<int64_t>(1), CreateScalar(false), {4, 3}, kInt64},
    ArgMinExtParams{{4, 2, 3}, kFloat16, CreateScalar<int64_t>(0), CreateScalar(false), {2, 3}, kInt64},
    ArgMinExtParams{{4, 2, 3}, kFloat16, CreateScalar<int64_t>(0), CreateScalar(true), {1, 2, 3}, kInt64}
));
}  // namespace ops
}  // namespace mindspore
